File: ios/Sources/NoiseProtocol/Crypto/Cipher.swift
import CryptoKit
import Foundation

/// ChaCha20-Poly1305 AEAD cipher for Noise protocol.
/// Nonce format: 4 zero bytes + 8 bytes little-endian counter.
public enum NoiseCipher {
    /// AEAD encrypt. Returns ciphertext || 16-byte tag.
    public static func encrypt(k: SymmetricKey, n: UInt64, ad: Data, plaintext: Data) throws -> Data {
        let nonce = try makeNonce(n)
        let sealedBox = try ChaChaPoly.seal(
            plaintext,
            using: k,
            nonce: nonce,
            authenticating: ad
        )
        return sealedBox.ciphertext + sealedBox.tag
    }

    /// AEAD decrypt. Throws on authentication failure.
    public static func decrypt(k: SymmetricKey, n: UInt64, ad: Data, ciphertext: Data) throws -> Data {
        guard ciphertext.count >= 16 else { throw NoiseError.decryptionFailed }
        let nonce = try makeNonce(n)
        let tagStart = ciphertext.count - 16
        let ct = ciphertext.prefix(tagStart)
        let tag = ciphertext.suffix(16)
        let sealedBox = try ChaChaPoly.SealedBox(
            nonce: nonce,
            ciphertext: ct,
            tag: tag
        )
        return try ChaChaPoly.open(sealedBox, using: k, authenticating: ad)
    }

    /// ChaChaPoly nonce: 4 zero bytes + 8 bytes little-endian n.
    private static func makeNonce(_ n: UInt64) throws -> ChaChaPoly.Nonce {
        var bytes = [UInt8](repeating: 0, count: 4)
        withUnsafeBytes(of: n.littleEndian) { bytes.append(contentsOf: $0) }
        return try ChaChaPoly.Nonce(data: bytes)
    }
}

File: android/lib/src/main/kotlin/com/noise/protocol/crypto/Cipher.kt
package com.noise.protocol.crypto

import java.nio.ByteBuffer
import java.nio.ByteOrder
import javax.crypto.Cipher
import javax.crypto.spec.GCMParameterSpec
import javax.crypto.spec.SecretKeySpec

import java.security.InvalidKeyException

/**
 * ChaCha20-Poly1305 AEAD cipher for Noise protocol.
 * Nonce format: 4 zero bytes + 8 bytes little-endian counter.
 */
object NoiseCipher {
    private val cipherLocal = ThreadLocal.withInitial {
        Cipher.getInstance("ChaCha20-Poly1305")
    }

    /** AEAD encrypt. Returns ciphertext || 16-byte tag. */
    fun encrypt(k: ByteArray, n: Long, ad: ByteArray, plaintext: ByteArray): ByteArray {
        val cipher = initCipher(Cipher.ENCRYPT_MODE, k, makeNonce(n))
        cipher.updateAAD(ad)
        return cipher.doFinal(plaintext)
    }

    /** AEAD decrypt. Throws on authentication failure. */
    fun decrypt(k: ByteArray, n: Long, ad: ByteArray, ciphertext: ByteArray): ByteArray {
        val cipher = initCipher(Cipher.DECRYPT_MODE, k, makeNonce(n))
        cipher.updateAAD(ad)
        return cipher.doFinal(ciphertext)
    }

    private fun initCipher(mode: Int, k: ByteArray, nonce: ByteArray): Cipher {
        val spec = javax.crypto.spec.IvParameterSpec(nonce)
        val key = SecretKeySpec(k, "ChaCha20-Poly1305")
        var cipher = cipherLocal.get()
        try {
            cipher.init(mode, key, spec)
        } catch (_: InvalidKeyException) {
            // JCA rejects same key+nonce on reused ChaCha20 instance (rekey scenario)
            cipher = Cipher.getInstance("ChaCha20-Poly1305")
            cipherLocal.set(cipher)
            cipher.init(mode, key, spec)
        }
        return cipher
    }

    /** ChaChaPoly nonce: 4 zero bytes + 8 bytes little-endian n. */
    private fun makeNonce(n: Long): ByteArray {
        val nonce = ByteArray(12)
        ByteBuffer.wrap(nonce, 4, 8)
            .order(ByteOrder.LITTLE_ENDIAN)
            .putLong(n)
        return nonce
    }
}

/**
 * AES-256-GCM AEAD cipher for Noise protocol.
 * Nonce format: 4 zero bytes + 8 bytes big-endian counter.
 */
object NoiseCipherAESGCM {
    private val cipherLocal = ThreadLocal.withInitial {
        Cipher.getInstance("AES/GCM/NoPadding")
    }

    /** AEAD encrypt. Returns ciphertext || 16-byte tag. */
    fun encrypt(k: ByteArray, n: Long, ad: ByteArray, plaintext: ByteArray): ByteArray {
        val cipher = cipherLocal.get()
        val nonce = makeNonce(n)
        cipher.init(Cipher.ENCRYPT_MODE, SecretKeySpec(k, "AES"), GCMParameterSpec(128, nonce))
        cipher.updateAAD(ad)
        return cipher.doFinal(plaintext)
    }

    /** AEAD decrypt. Throws on authentication failure. */
    fun decrypt(k: ByteArray, n: Long, ad: ByteArray, ciphertext: ByteArray): ByteArray {
        val cipher = cipherLocal.get()
        val nonce = makeNonce(n)
        cipher.init(Cipher.DECRYPT_MODE, SecretKeySpec(k, "AES"), GCMParameterSpec(128, nonce))
        cipher.updateAAD(ad)
        return cipher.doFinal(ciphertext)
    }

    /** AESGCM nonce: 4 zero bytes + 8 bytes big-endian n. */
    private fun makeNonce(n: Long): ByteArray {
        val nonce = ByteArray(12)
        ByteBuffer.wrap(nonce, 4, 8).order(ByteOrder.BIG_ENDIAN).putLong(n)
        return nonce
    }
}
